src/main/scala/is/hail/expr/ir/functions/Functions.scala

```diff
@@ -7,38 +7,60 @@ import is.hail.utils._
 import is.hail.asm4s.coerce
 
 import scala.collection.mutable
-import scala.reflect.ClassTag
 
 object IRFunctionRegistry {
 
-  val registry: mutable.Map[String, Seq[(Seq[Type], Seq[IR] => IR)]] = mutable.Map().withDefaultValue(Seq.empty)
+  val irRegistry: mutable.Map[String, Seq[(Seq[Type], Seq[IR] => IR)]] = mutable.Map().withDefaultValue(Seq.empty)
+
+  val codeRegistry: mutable.Map[String, Seq[(Seq[Type], IRFunction)]] = mutable.Map().withDefaultValue(Seq.empty)
 
   def addIRFunction(f: IRFunction) {
-    val l = registry(f.name)
-    registry.put(f.name,
-      l :+ (f.argTypes, { args: Seq[IR] =>
-        ApplyFunction(f, args)
-      }))
+    val l = codeRegistry(f.name)
+    codeRegistry.put(f.name,
+      l :+ (f.argTypes, f))
   }
 
   def addIR(name: String, types: Seq[Type], f: Seq[IR] => IR) {
-    val l = registry(name)
-    registry.put(name, l :+ ((types, f)))
+    val l = irRegistry(name)
+    irRegistry.put(name, l :+ ((types, f)))
+  }
+
+  def lookupFunction(name: String, args: Seq[Type]): Option[IRFunction] = {
+    val validF = codeRegistry(name).flatMap { case (ts, f) =>
+      if (ts.length == args.length) {
+        ts.foreach(_.clear())
+        if ((ts, args).zipped.forall(_.unify(_)))
+          Some(f)
+        else
+          None
+      } else
+        None
+    }
+
+    validF match {
+      case Seq() => None
+      case Seq(x) => Some(x)
+      case _ => fatal(s"Multiple IRFunctions found that satisfy $name$args.")
+    }
   }
 
-  def lookupFunction(name: String, args: Seq[Type]): Option[Seq[IR] => IR] = {
+  def lookupConversion(name: String, args: Seq[Type]): Option[Seq[IR] => IR] = {
```

Comment:
`lookupConversion` and `lookupFunction` are suspiciously similar. I'm not sure I see the compelling reason to keep them separate.

Would the code look simpler if we had one `MultiMap[String, IRFun]` where `IRFun` was something like:

```scala
trait IRFun {
  def args: Seq[Type]
  def ret: Type
  def emit(
    emitter: IR => (Code[_], Code[Boolean], Code[_]),
    mb: MethodBuilder,
    args: IR*
  ): (Code[Unit], Code[Boolean], Code[_])
}

class IRMacro(
  val args: Seq[Type],
  val ret: Type,
  val impl: Array[IR] => IR
) extends IRFun {
  def emit(
    emitter: IR => (Code[_], Code[Boolean], Code[_]),
    mb: MethodBuilder,
    args: IR*
  ): (Code[Unit], Code[Boolean], Code[_]) = emitter(impl(args.toArray))
}

class JVMFun(
  val args: Seq[Type],
  retType: Type,
  impl: (MethodBuilder, Array[Code[_]]) => Code[_]
) extends IRFun {
  def emit(
    emitter: IR => (Code[_], Code[Boolean], Code[_]),
    mb: MethodBuilder,
    args: IR*
  ): (Code[Unit], Code[Boolean], Code[_]) = {
    val (setup, ms, vs) = args.map(emitter).unzip
    (setup, ms.fold(_ || _), impl(mb, vs))
  }
}
```

Reply:
Or maybe let JVMFun also control missingness?

```scala
class MissingAwareJVMFun(
  val args: Seq[Type],
  retType: Type,
  impl: (MethodBuilder, Array[Code[_]]) => (Code[Unit], Code[Boolean], Code[_])
) extends IRFun {
  def emit(
    emitter: IR => (Code[Unit], Code[Boolean], Code[_]),
    mb: MethodBuilder,
    args: IR*
  ): (Code[Unit], Code[Boolean], Code[_]) = {
    val (setups, ms, vs) = args.map(emitter).unzip
    val (setup, m, v) = impl(mb, vs)
    (Code(setups ++ setup:_*), (ms :+ m).fold(_ || _), v)
  }
}

```

Reply:
I don't really want to entangle them because of two resasons:
- Things that can be translated directly (`Seq[IR] => IR`) are maybe things that we don't want to keep in the function registry once we can directly pass AST nodes from Python. I kind of want all of them to go away.
- Even if we do want to keep them in the function registry, I want to be able to look up the non-IR conversions post optimization, when we go to compile the code. Once we're no longer translating from AST, we don't need to store the implementation as part of the IR node and can look it up after we're done optimizing the overall IR. I feel like the things that are implemented in terms of existing IR nodes should just go directly into the IR tree and optimized away with the rest of it. (this is kind of what currently happens; `Infer` on `Apply` nodes does a lookup on the IRFunction registry to be able to properly infer the return type)